File: app/src/main/java/org/sil/storyproducer/controller/remote/WholeStoryBackTranslationActivity.java
package org.sil.storyproducer.controller.remote;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.graphics.Bitmap;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.util.DisplayMetrics;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.SeekBar;
import android.widget.Switch;

import org.sil.storyproducer.R;
import org.sil.storyproducer.controller.phase.PhaseBaseActivity;
import org.sil.storyproducer.model.StoryState;
import org.sil.storyproducer.tools.BitmapScaler;
import org.sil.storyproducer.tools.file.AudioFiles;
import org.sil.storyproducer.tools.file.FileSystem;
import org.sil.storyproducer.tools.file.ImageFiles;
import org.sil.storyproducer.tools.media.AudioPlayer;
import org.sil.storyproducer.tools.media.MediaHelper;
import org.sil.storyproducer.tools.toolbar.RecordingToolbar;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by annmcostantino on 1/14/2018.
 */

public class WholeStoryBackTranslationActivity extends PhaseBaseActivity {

    private final static float BACKGROUND_VOLUME = 0.0f;        //makes for no background music but still keeps the functionality in there if we decide to change it later

    private RelativeLayout rootView;
    private ImageView wStoryImageView;
    private ImageButton playButton;
    private SeekBar videoSeekBar;
    private AudioPlayer narrationPlayer;
    private AudioPlayer backgroundPlayer;
    private boolean backgroundAudioExists;

    private int slideNumber = 0;
    private int CONTENT_SLIDE_COUNT = 0;
    private String storyName;
    private boolean isVolumeOn = true;
    private List<Integer> backgroundAudioJumps;

    //recording toolbar vars
    private String recordFilePath;
    private RecordingToolbar recordingToolbar;

    private boolean isFirstTime = true;         //used to know if it is the first time the activity is started up for playing the vid
    private int startPos = -1;
    private long startTime = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_whole_story);
        rootView = findViewById(R.id.phase_frame);

        //get the story name
        storyName = StoryState.getStoryName();
        CONTENT_SLIDE_COUNT = FileSystem.getContentSlideAmount(storyName);

        //get the ui
        wStoryImageView = findViewById(R.id.wholeStoryImageView);
        playButton = findViewById(R.id.playButton);
        videoSeekBar = findViewById(R.id.videoSeekBar);

        setBackgroundAudioJumps();

        setSeekBarListener();

        setPic(wStoryImageView);     //set the first image to show

        //set the recording toolbar stuffs
        recordFilePath = AudioFiles.getWholeStory(StoryState.getStoryName()).getPath();
        View rootViewToolbar = getLayoutInflater().inflate(R.layout.toolbar_for_recording, rootView, false);
        setToolbar(rootViewToolbar);
        invalidateOptionsMenu();
        setVolumeSwitchAndFloatingButtonVisible();
        recordingToolbar.keepToolbarVisible();
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        MenuItem item =  menu.getItem(0);
        item.setIcon(R.drawable.ic_wholestory);
        return true;
    }

    /**
     * Starts the background music player
     */
    private void playBackgroundMusic() {
        if (backgroundAudioExists) {
            backgroundPlayer.playAudio();
        }
    }

    /**
     * Sets the array list for all the jump points that the background music has to make
     */
    private void setBackgroundAudioJumps() {
        int audioStartValue = 0;
        backgroundAudioJumps = new ArrayList<>();
        backgroundAudioJumps.add(0, audioStartValue);
        for(int k = 0; k < CONTENT_SLIDE_COUNT; k++) {
            String lwcPath = AudioFiles.getLWC(storyName, k).getPath();
            audioStartValue += MediaHelper.getAudioDuration(lwcPath) / 1000;
            backgroundAudioJumps.add(k, audioStartValue);
        }
        backgroundAudioJumps.add(audioStartValue);        //this last one is just added for the copyrights slide
    }

    public void onStart() {
        super.onStart();
        //create audio players
        narrationPlayer = new AudioPlayer();
        narrationPlayer.onPlayBackStop(new MediaPlayer.OnCompletionListener() {
            @Override
            public void onCompletion(MediaPlayer mp) {
                slideNumber++;         //move to the next slide
                if(slideNumber < CONTENT_SLIDE_COUNT) {     //not at the end of video
                    playVideo();
                } else {                            //at the end of video so special case
                    //makeLogIfNecessary(true);

                    videoSeekBar.setProgress(CONTENT_SLIDE_COUNT);
                    playButton.setImageResource(R.drawable.ic_play_gray);
                    setPic(wStoryImageView);     //sets the pic to the end image
                }
            }
        });
        //recordingToolbar.hideFloatingActionButton();

        backgroundPlayer = new AudioPlayer();
        backgroundPlayer.setVolume(BACKGROUND_VOLUME);
        File backgroundAudioFile = AudioFiles.getSoundtrack(StoryState.getStoryName());
        if (backgroundAudioFile.exists()) {
            backgroundAudioExists = true;
            backgroundPlayer.setPath(backgroundAudioFile.getPath());
        } else {
            backgroundAudioExists = false;
        }
    }

    /*private void markLogStart() {
        startPos = slideNumber;
        startTime = System.currentTimeMillis();
    }

    private void makeLogIfNecessary(){
        makeLogIfNecessary(false);
    }

    private void makeLogIfNecessary(boolean request){
        if(narrationPlayer.isAudioPlaying() || backgroundPlayer.isAudioPlaying()
                || request){
            if(startPos!=-1) {
                WholeStoryEntry.saveFilteredLogEntry(startPos, slideNumber,
                        System.currentTimeMillis() - startTime);
                startPos=-1;
            }
        }
    }*/

    @Override
    public void onPause() {
        super.onPause();
        pauseVideo();
        if (recordingToolbar != null) {
            recordingToolbar.onClose();
            recordingToolbar.closeToolbar();
        }
    }

    @Override
    public void onResume() {
            super.onResume();
            //recordingToolbar.hideFloatingActionButton();
    }

    @Override
    public void onStop() {
        super.onStop();
        narrationPlayer.release();
        backgroundPlayer.release();
        if (recordingToolbar != null) {
            recordingToolbar.onClose();
            recordingToolbar.closeToolbar();
            recordingToolbar.releaseToolbarAudio();
        }

    }

    /**
     * Plays the video and runs every time the audio is completed
     */
    void playVideo() {
        setPic(wStoryImageView);                                                             //set the next image
        File audioFile = AudioFiles.getDraft(storyName, slideNumber);
        //set the next audio
        if (audioFile.exists()) {
            narrationPlayer.setVolume((isVolumeOn)? 1.0f : 0.0f); //set the volume on or off based on the boolean
            narrationPlayer.setPath(audioFile.getPath());
            narrationPlayer.playAudio();
        }

        videoSeekBar.setProgress(slideNumber);
    }

    /**
     * Button action for playing/pausing the audio
     * @param view button to set listeners for
     */
    public void onClickPlayPauseButton(View view) {
        if(narrationPlayer.isAudioPlaying()) {
            pauseVideo();

        }
        //if no draft audio exists
        else if(!AudioFiles.allDraftsComplete(storyName,FileSystem.getContentSlideAmount(storyName))){
            AlertDialog dialog = new AlertDialog.Builder(this)
                    .setTitle(this.getString(R.string.wsbt_alert_title))
                    .setMessage(this.getString(R.string.wsbt_alert_text))
                    .setPositiveButton(this.getString(R.string.ok), new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int id) {

                        }
                    }).create();
            dialog.show();
        } else {
            //markLogStart();

            playButton.setImageResource(R.drawable.ic_pause_gray);

            if(slideNumber >= CONTENT_SLIDE_COUNT) {        //reset the video to the beginning because they already finished it
                videoSeekBar.setProgress(0);
                slideNumber = 0;
                playBackgroundMusic();
                playVideo();
            } else {
                resumeVideo();
            }
        }
    }

    /**
     * helper function for pausing the video
     */
    private void pauseVideo() {
        //makeLogIfNecessary();
        narrationPlayer.pauseAudio();
        backgroundPlayer.pauseAudio();
        playButton.setImageResource(R.drawable.ic_play_gray);
    }

    /**
     * helper function for resuming the video
     */
    private void resumeVideo() {
        if(isFirstTime) {           //actually start playing the video if playVideo() has never been called
            playVideo();
            isFirstTime = false;
        } else {
            narrationPlayer.resumeAudio();
            if (backgroundAudioExists) {
                backgroundPlayer.resumeAudio();
            }
        }
    }

    /**
     * Sets the seekBar listener for the video seek bar
     */
    private void setSeekBarListener() {
        videoSeekBar.setMax(CONTENT_SLIDE_COUNT);      //set the progress bar to have as many markers as images
        videoSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onStopTrackingTouch(SeekBar sBar){
            }
            @Override
            public void onStartTrackingTouch(SeekBar sBar){
            }
            @Override
            public void onProgressChanged(SeekBar sBar, int progress, boolean fromUser) {
                if(fromUser) {
                    //makeLogIfNecessary();

                    slideNumber = progress;
                    narrationPlayer.stopAudio();
                    if(backgroundAudioExists) {
                        backgroundPlayer.seekTo(backgroundAudioJumps.get(slideNumber));
                        if (!backgroundPlayer.isAudioPlaying()) {
                            backgroundPlayer.resumeAudio();
                        }
                    }
                    if(slideNumber == CONTENT_SLIDE_COUNT) {
                        playButton.setImageResource(R.drawable.ic_play_gray);
                        setPic(wStoryImageView);     //sets the pic to the end image
                    } else {
                        //markLogStart();
                        playVideo();
                        playButton.setImageResource(R.drawable.ic_pause_gray);
                    }

                }
            }
        });
    }

    /**
     * helper function that resets the video to the beginning and turns off the sound
     */
    private void resetVideoWithSoundOff() {

        videoSeekBar.setProgress(0);
        slideNumber = 0;
        narrationPlayer.setVolume(0.0f);
        Switch volumeSwitch = findViewById(R.id.volumeSwitch);
        backgroundPlayer.stopAudio();
        volumeSwitch.setChecked(false);
        backgroundPlayer.stopAudio();
        backgroundPlayer.setVolume(0.0f);
        playBackgroundMusic();
        isVolumeOn = false;

       //markLogStart();
        if(AudioFiles.allDraftsComplete(storyName,FileSystem.getContentSlideAmount(storyName))){
            playButton.setImageResource(R.drawable.ic_pause_gray);
            playVideo();
        }
    }

    /**
     * Makes the volume switch visible so it can be used
     */
    private void setVolumeSwitchAndFloatingButtonVisible() {
        //make the floating button visible
        //recordingToolbar.showFloatingActionButton();
        //make the sounds stuff visible
        ImageView soundOff = findViewById(R.id.soundOff);
        ImageView soundOn = findViewById(R.id.soundOn);
        Switch volumeSwitch = findViewById(R.id.volumeSwitch);
        soundOff.setVisibility(View.VISIBLE);
        soundOn.setVisibility(View.VISIBLE);
        volumeSwitch.setVisibility(View.VISIBLE);
        //set the volume switch change listener
        volumeSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if(isChecked) {
                    narrationPlayer.setVolume(1.0f);
                    backgroundPlayer.setVolume(BACKGROUND_VOLUME);
                    isVolumeOn = true;
                } else {
                    narrationPlayer.setVolume(0.0f);
                    backgroundPlayer.setVolume(0.0f);
                    isVolumeOn = false;
                }
            }
        });
    }

    /**
     * This function allows the picture to scale with the phone's screen size.
     *
     * @param aView    The ImageView that will contain the picture.
     */
    private void setPic(View aView) {
        if (aView == null || !(aView instanceof ImageView)) {
            return;
        }

        ImageView slideImage = (ImageView) aView;
        Bitmap slidePicture = ImageFiles.getBitmap(storyName, slideNumber);
        if(slideNumber == CONTENT_SLIDE_COUNT) {                //gets the end image if we are at the end of the story
            slidePicture = ImageFiles.getBitmap(storyName, ImageFiles.COPYRIGHT);
        }

        if(slidePicture == null){
            Snackbar.make(rootView, "Could Not Find Picture", Snackbar.LENGTH_SHORT).show();
        }

        //Get the height of the phone.
        DisplayMetrics phoneProperties = getResources().getDisplayMetrics();
        int height = phoneProperties.heightPixels;
        double scalingFactor = 0.4;
        height = (int)(height * scalingFactor);

        //scale bitmap
        slidePicture = BitmapScaler.scaleToFitHeight(slidePicture, height);

        //Set the height of the image view
        slideImage.getLayoutParams().height = height;
        slideImage.requestLayout();

        slideImage.setImageBitmap(slidePicture);
    }

    /**
     * Initializes the toolbar and toolbar buttons.
     */
    private void setToolbar(View toolbar){
        recordingToolbar = new RecordingToolbar(this, toolbar, rootView, true, false, false, true, recordFilePath, recordFilePath, null, new RecordingToolbar.RecordingListener() {
            @Override
            public void onStoppedRecording() {

            }
            @Override
            public void onStartedRecordingOrPlayback(boolean isRecording) {
                resetVideoWithSoundOff();
            }
        });
        //recordingToolbar.hideFloatingActionButton();
        //The following allows for a touch from user to close the toolbar and make the fab visible.
        //This does not stop the recording
        /*RelativeLayout dummyView = (RelativeLayout) rootView.findViewById(R.id.activity_wholestorybacktranslation);
        dummyView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (recordingToolbar != null && recordingToolbar.isOpen() && !recordingToolbar.isRecording()) {
                    recordingToolbar.keepToolbarVisible();
                    recordingToolbar.hideFloatingActionButton();
                }
            }
        });*/
    }



}

File: app/src/main/java/org/sil/storyproducer/controller/remote/WholeStoryBackTranslationActivity.kt
// TODO @pwhite: WholeStoryBackTranslationActivity and LearnActivity are
// extremely similar. The latter allows viewing of a *template*, and the former
// allows viewing of a *story*, which is essentially a translated template. The
// major difference is that the backtranslation should also allow uploading,
// but this is does not prevent us from extracting the common functionality.
package org.sil.storyproducer.controller.remote

import android.content.Context
import android.media.MediaPlayer
import android.os.Bundle
import android.provider.Settings
import android.support.graphics.drawable.VectorDrawableCompat
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.android.volley.VolleyError
import com.android.volley.Response;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.NetworkResponse;
import org.apache.commons.io.IOUtils
import org.json.JSONObject
import org.json.JSONException
import org.sil.storyproducer.BuildConfig
import org.sil.storyproducer.R
import org.sil.storyproducer.controller.phase.PhaseBaseActivity
import org.sil.storyproducer.model.SLIDE_NUM
import org.sil.storyproducer.model.SlideType
import org.sil.storyproducer.model.Recording
import org.sil.storyproducer.model.UploadState
import org.sil.storyproducer.model.Workspace
import org.sil.storyproducer.tools.Network.VolleySingleton
import org.sil.storyproducer.tools.Network.paramStringRequest
import org.sil.storyproducer.tools.file.getStoryChildInputStream
import org.sil.storyproducer.tools.file.getStoryUri
import org.sil.storyproducer.tools.media.AudioPlayer
import org.sil.storyproducer.tools.media.MediaHelper
import org.sil.storyproducer.tools.toolbar.PlayBackRecordingToolbar
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.text.Charsets.UTF_8

fun getPhoneId(context: Context): String {
    return Settings.Secure.getString(context.applicationContext.contentResolver, Settings.Secure.ANDROID_ID)
}

fun sendProjectSpecificRequest(
    context: Context,
    relativeUrl: String,
    onSuccess: (JSONObject) -> Unit,
    onFailure: (VolleyError) -> Unit,
    params: HashMap<String, String> = HashMap()) {

    params["Key"] = context.getString(R.string.api_token)
    params["PhoneId"] = getPhoneId(context)
    val url = Workspace.getRoccUrlPrefix(context) + relativeUrl
    val req = object : StringRequest(Method.POST, url, {
        Log.i("LOG_VOLLEY", it)
        var jsonObject: JSONObject? = null
        try {
            jsonObject = JSONObject(it)
        } catch (e: JSONException) {
            Toast.makeText(context, "The request was successful, but the response was of an unexpected form.", Toast.LENGTH_SHORT).show()
        }
        if (jsonObject != null) {
            onSuccess(jsonObject)
        }
    }, {
        Log.e("LOG_VOLLEY", "HIT ERROR")
        Log.e("LOG_VOLLEY", it.toString())
        val nr = it.networkResponse
        if (nr != null) {
            Toast.makeText(context, "${nr.statusCode}: ${String(nr.data, UTF_8)}", Toast.LENGTH_SHORT).show()
        } else {
            Toast.makeText(context, "Failed to connect to server.", Toast.LENGTH_SHORT).show()
        }
        onFailure(it)
    }) {
        override fun getParams(): Map<String, String> {
            return params
        }
    }
    VolleySingleton.getInstance(context.applicationContext).addToRequestQueue(req)
}

// TODO @pwhite: Perhaps this function is more appropriate in another file, but
// hey, code is code, and it semantically doesn't matter where we put it. It
// matters for discoverability and convenience though, so worth thinking about
// once it is used more.
fun sendSlideSpecificRequest(
    context: Context,
    slideNumber: Int,
    relativeUrl: String,
    content: String,
    onSuccess: (JSONObject) -> Unit,
    onFailure: (VolleyError) -> Unit,
    js: HashMap<String, String> = HashMap()) {

    if (Workspace.activeStory.remoteId != null) {
        js["StoryId"] = Workspace.activeStory.remoteId.toString()
    }
    js["TemplateTitle"] = Workspace.activeStory.title
    js["SlideNumber"] = slideNumber.toString()
    js["Data"] = content
    sendProjectSpecificRequest(context, relativeUrl, {
        val newStoryId = it.getInt("StoryId")
        Log.e("@pwhite", "Received id $newStoryId")
        if (Workspace.activeStory.remoteId == null) {
          Log.i("@pwhite", "Setting active story id from null to $newStoryId")
          Workspace.activeStory.remoteId = newStoryId
        } else {
          Log.e("SanityCheck", "Response id ($newStoryId) should be the same story id as stored (${Workspace.activeStory.remoteId})")
        }
        onSuccess(it)
    }, onFailure, js)
}

/**
 * Created by annmcostantino on 1/14/2018.
 *
 * An interface for doing back translations on the whole story. There is an image and a seekbar
 * which provide a UI for watching the video with both slides and audio. There is also a recording
 * toolbar for recording and uploading audio. It is us
 */


class WholeStoryBackTranslationFragment : Fragment(), PlayBackRecordingToolbar.ToolbarMediaListener {

    class DraftSlide(val slideNumber: Int, val duration: Int, val startTime: Int, val filename: String) {}

    private lateinit var wholeStoryImageView: ImageView
    private lateinit var playButton: ImageButton
    private lateinit var seekBar: SeekBar

    private var mSeekBarTimer = Timer()
    private var draftPlayer: AudioPlayer = AudioPlayer()
    private var seekbarStartTime: Long = -1

    private var isVolumeOn = true
    private var viewIsCreated = false

    private var recordingToolbar: PlayBackRecordingToolbar = PlayBackRecordingToolbar()
    private lateinit var uploadAudioButtonManager: UploadAudioButtonManager

    private var currentSlideIndex: Int = 0
    private val translatedSlides: MutableList<DraftSlide> = ArrayList()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        val rootView = inflater.inflate(R.layout.activity_whole_story, container, false)

        val bundle = Bundle()
        bundle.putInt(SLIDE_NUM, 0)
        recordingToolbar.arguments = bundle
        childFragmentManager.beginTransaction().replace(R.id.toolbar_for_recording_toolbar, recordingToolbar).commit()
        recordingToolbar.keepToolbarVisible()

        wholeStoryImageView = rootView.findViewById(R.id.fragment_image_view)
        playButton = rootView.findViewById(R.id.fragment_reference_audio_button)
        seekBar = rootView.findViewById(R.id.videoSeekBar)

        playButton.setOnClickListener {
            if (draftPlayer.isAudioPlaying) {
                pauseStoryAudio()
            } else {
                // If the video is basically already finished, restart it.
                if (seekBar.progress >= seekBar.max - 100) {
                    seekBar.progress = 0
                }
                playStoryAudio()
            }
        }

        uploadAudioButtonManager = UploadAudioButtonManager(
            context!!,
            rootView.findViewById(R.id.upload_audio_botton),
            { Workspace.activeStory.wholeStoryBackTranslationUploadState },
            { Workspace.activeStory.wholeStoryBackTranslationUploadState = it },
            { Workspace.activeStory.wholeStoryBackTAudioFile }, 
            null)

        seekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            var wasPlayingBeforeTouch = false
            override fun onStopTrackingTouch(sBar: SeekBar) {
                if (wasPlayingBeforeTouch) {
                    // Always start at the beginning of the slide.
                    if (currentSlideIndex < translatedSlides.size) {
                        seekBar.progress = translatedSlides[currentSlideIndex].startTime
                    }
                    playStoryAudio()
                }
            }

            override fun onStartTrackingTouch(sBar: SeekBar) {
                wasPlayingBeforeTouch = draftPlayer.isAudioPlaying
            }

            override fun onProgressChanged(sBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    setSlideFromSeekbar()
                }
            }
        })

        val volumeSwitch = rootView.findViewById<Switch>(R.id.volumeSwitch)
        volumeSwitch.setOnCheckedChangeListener { _, isChecked ->
            isVolumeOn = if (isChecked) {
                draftPlayer.setVolume(1.0f)
                true
            } else {
                draftPlayer.setVolume(0.0f)
                false
            }
        }

        // Compute story audio duration
        var lastEndTime = 0
        Workspace.activeStory.slides.forEachIndexed { slideNumber, slide ->
            // Don't play the copyright slides.
            if (slide.slideType == SlideType.FRONTCOVER || slide.slideType == SlideType.NUMBEREDPAGE) {
                val filename = slide.draftRecordings.selectedFile?.fileName
                if (filename != null) {
                    val duration = (MediaHelper.getAudioDuration(context!!, getStoryUri(filename)!!) / 1000).toInt()
                    val startTime = lastEndTime
                    lastEndTime = startTime + duration
                    translatedSlides.add(DraftSlide(slideNumber, duration, startTime, filename))
                }
            }
        }

        seekBar.max = if (translatedSlides.isNotEmpty()) {
            val lastSlide = translatedSlides.last()
            lastSlide.startTime + lastSlide.duration
        } else {
            0
        }
        seekBar.progress = 0
        setSlideFromSeekbar()

        viewIsCreated = true

        return rootView
    }

    override fun onPause() {
        super.onPause()
        draftPlayer.release()
    }

    override fun onResume() {
        super.onResume()
        draftPlayer = AudioPlayer()
        draftPlayer.onPlayBackStop(MediaPlayer.OnCompletionListener {
            if (draftPlayer.isAudioPrepared) {
                // If the video has reached the end, then pause; otherwise,
                // just play the next slide.
                if (currentSlideIndex >= translatedSlides.size - 1) {
                    pauseStoryAudio()
                } else {
                    seekBar.progress = translatedSlides[currentSlideIndex + 1].startTime
                    playStoryAudio()
                }
            }
        })

        val currentActivity = activity!!
        mSeekBarTimer = Timer()
        mSeekBarTimer.schedule(object : TimerTask() {
            override fun run() {
                currentActivity.runOnUiThread {
                    if (recordingToolbar.isRecording) {
                        setSlideFromSeekbar()
                    }
                    if (recordingToolbar.isRecording || recordingToolbar.isAudioPlaying) {
                        seekBar.progress = minOf((System.currentTimeMillis() - seekbarStartTime).toInt(), seekBar.max)
                    } else if (draftPlayer.isAudioPrepared) {
                        seekBar.progress = translatedSlides[currentSlideIndex].startTime + draftPlayer.currentPosition
                    } else {
                        seekBar.progress = 0
                    }
                }
            }
        }, 0, 33)
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (viewIsCreated && !isVisibleToUser) {
            pauseStoryAudio()
        }
    }

    private fun setSlideFromSeekbar() {
        if (translatedSlides.isNotEmpty()) {
            val time = seekBar.progress
            var slideIndexBeforeSeekBar = translatedSlides.indexOfLast { it.startTime <= time }
            if (slideIndexBeforeSeekBar != currentSlideIndex || !draftPlayer.isAudioPrepared) {
                currentSlideIndex = slideIndexBeforeSeekBar
                val slide = translatedSlides[currentSlideIndex]
                PhaseBaseActivity.setPic(context!!, wholeStoryImageView, slide.slideNumber)
                draftPlayer.setStorySource(context!!, slide.filename)
            }
        }
    }

    override fun onStoppedToolbarRecording() {
        super.onStoppedToolbarRecording()
        Workspace.activeStory.wholeStoryBackTranslationUploadState = UploadState.NOT_UPLOADED
        uploadAudioButtonManager.refreshBackground()
    }

    override fun onStoppedToolbarMedia() {
        seekBar.progress = 0
        setSlideFromSeekbar()
    }

    override fun onStartedToolbarMedia() {
        pauseStoryAudio()
        seekBar.progress = 0
        currentSlideIndex = 0
        //This gets the progress bar to show the right time.
        seekbarStartTime = System.currentTimeMillis()
    }

    /**
     * Plays the audio
     */
    internal fun playStoryAudio() {
        recordingToolbar.stopToolbarMedia()
        setSlideFromSeekbar()
        draftPlayer.pauseAudio()
        seekbarStartTime = System.currentTimeMillis()
        draftPlayer.setVolume(if (isVolumeOn) 1.0f else 0.0f) //set the volume on or off based on the boolean
        draftPlayer.playAudio()
        playButton.setImageResource(R.drawable.ic_pause_white_48dp)
    }

    /**
     * helper function for pausing the video
     */
    private fun pauseStoryAudio() {
        draftPlayer.pauseAudio()
        playButton.setImageResource(R.drawable.ic_play_arrow_white_48dp)
    }
}

